arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -48,17 +48,40 @@ SECTION_FUNC(TEXT, z_arm_cpu_idle_init)
 #endif
 	bx	lr
 
-SECTION_FUNC(TEXT, arch_cpu_idle)
-#ifdef CONFIG_TRACING
+.macro _prologue
+#if defined(CONFIG_TRACING) || defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
 	push	{r0, lr}
+#if defined(CONFIG_TRACING)
 	bl	sys_trace_idle
+#endif
+#if defined(CONFIG_ARM_ON_ENTER_CPU_IDLE_HOOK)
+	bl	z_arm_on_enter_cpu_idle
+	/* Preserve the return value so that it can be used later,
+	 * see the _sleep_if_allowed macro.
+	 */
+	mov	r2, r0
```

Comment:
why are you splitting this into `prologue` and `_sleep_if_allowed`? Can't we have just a single entry-point for this?

Reply:
I initially tried to reuse the pushing and popping code that was already present for tracing. And I unnecessarily got fixed on this approach.
I reworked it. Thanks for the hint and please take another look.